### PR TITLE
[spirv] Translate non-entry functions and function calls

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: tools/clang/include/clang/SPIRV/DeclResultIdMapper.h [hunk 1]
```diff
@@ -59,22 +59,29 @@ class DeclResultIdMapper {
...
+  ///
+  /// This method will panic if the given decl is not registered.
   uint32_t getDeclResultId(const NamedDecl *decl) const;
 
-  /// \brief Returns the <result-id> for the given remapped Decl. Returns zero
-  /// if it is not a registered remapped Decl.
+  /// \brief Returns the <result-id> for the given decl if already registered;
+  /// otherwise, treats the given decl as a normal decl and returns a newly
+  /// assigned <result-id> for it.
+  uint32_t getOrRegDeclResultId(const NamedDecl *decl);
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/DeclResultIdMapper.h]: I would rename this to `getOrRegisterDeclResultId`

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 2]
```diff
@@ -833,6 +908,11 @@ case BO_##kind : {                                                             \
   DeclResultIdMapper declIdMapper;
   TypeTranslator typeTranslator;
 
+  /// A queue of decls reachable from the entry function. Decls inserted into
+  /// this queue will persist to avoid duplicated translations. And we'd like
+  /// a deterministic order of iterating the queue for finding the next decl
+  /// to translate. So we need SetVector here.
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: +1

File: tools/clang/test/CodeGenSPIRV/fn.call.hlsl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// CHECK-NOT: %fnNoCaller = OpFunction
+void fnNoCaller() {}
+
+int fnOneParm(int v) { return v; }
+
+int fnTwoParm(int m, int n) { return m + n; }
+
+int fnCallOthers(int v) { return fnOneParm(v); }
+
+// Recursive functions are disallowed by the front end.
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/fn.call.hlsl]: Thanks for this comment! I did not know :)

File: docs/SPIR-V.rst [hunk 4]
```diff
@@ -260,6 +260,51 @@ Control flows
...
 
+Functions
+---------
+
+All functions reachable from the entry-point function will be translated into SPIR-V code. Functions not reachable from the entry-point function will be ignored.
+
+Function parameter
+++++++++++++++++++
+
+For a function ``f`` who has a parameter of type ``T``, the generated SPIR-V signature will use type ``T*`` for the parameter. At every call site of ``f``, additional local variables will be allocated to hold the actual arguments. The local variables are passed in as direct function arguments. For example::
```
Comment (antiagainst) [inline docs/SPIR-V.rst]: Done

File: docs/SPIR-V.rst [hunk 5]
```diff
@@ -260,6 +260,51 @@ Control flows
...
+                ...
+
+           %f = OpFunction %v4float None %fnType
+           %a = OpFunctionParameter %f32PtrType
+           %b = OpFunctionParameter %i32PtrType
+                ...
+
+      %caller = OpFunction ...
+                ...
+     %aAlloca = OpVariable %_ptr_Function_float Function
```
Comment (antiagainst) [inline docs/SPIR-V.rst]: `%_ptr_Function_float` is the default name generated from SPIRV-Tools. We should keep it for consistency.

File: docs/SPIR-V.rst [hunk 6]
```diff
@@ -260,6 +260,51 @@ Control flows
...
+     %bAlloca = OpVariable %_ptr_Function_int Function
```
Comment (antiagainst) [inline docs/SPIR-V.rst]: Same

**Pull request merged**